nestkernel/secondary_event.h

```diff
@@ -0,0 +1,449 @@
+    left_to_copy -= sizeof( unsigned int );
+  }
+
+  pos += num_uints;
+}
+
+/**
+ * Template class for the storage and communication of a std::vector of type
+ * DataType. The class provides the functionality to communicate homogeneous
+ * data of type DataType. The second template type Subclass (which should be
+ * chosen as the derived class itself) is used to distinguish derived classes
+ * with the same DataType. This is required because of the included static
+ * variables in the base class (as otherwise all derived classes with the same
+ * DataType would share the same static variables).
+ *
+ * Technically the DataSecondaryEvent only contains iterators pointing to
+ * the memory location of the std::vector< DataType >.
+ *
+ * Conceptually, there is a one-to-one mapping between a SecondaryEvent
+ * and a SecondaryConnectorModel. The synindex of this particular
+ * SecondaryConnectorModel is stored as first element in the static vector
+ * supported_syn_ids_ on model registration. There are however reasons (e.g.
+ * the usage of CopyModel or the creation of the labeled synapse model
+ * duplicates for pyNN) which make it necessary to register several
+ * SecondaryConnectorModels with one SecondaryEvent. Therefore the synindices
+ * of all these models are added to supported_syn_ids_. The
+ * supports_syn_id()-function allows testing if a particular synid is mapped
+ * with the SecondaryEvent in question.
+ */
+template < typename DataType, typename Subclass >
+class DataSecondaryEvent : public SecondaryEvent
+{
+private:
+  // we chose std::vector over std::set because we expect this to be short
+  static std::vector< synindex > supported_syn_ids_;
+  static size_t coeff_length_; // length of coeffarray
+
+  union CoeffarrayBegin
+  {
+    std::vector< unsigned int >::iterator as_uint;
+    typename std::vector< DataType >::iterator as_DataType;
+
+    CoeffarrayBegin() {}; // need to provide default constructor due to
+                          // non-trivial constructors of iterators
+  } coeffarray_begin_;
+
+  union CoeffarrayEnd
+  {
+    std::vector< unsigned int >::iterator as_uint;
+    typename std::vector< DataType >::iterator as_DataType;
+
+    CoeffarrayEnd() {}; // need to provide default constructor due to
+                        // non-trivial constructors of iterators
+  } coeffarray_end_;
+
+public:
+  /**
+   * This function adds the ids of connection models that use this
+   * event type. This is called when the model is registered with the
+   * kernel and when the corresponded connector model is copied.
+   *
+   * This function needs to be a virtual function of the base class
+   * DataSecondaryEvent as it is usually called from a pointer or
+   * reference of type SecondaryEvent.
+   *
+   * See also:
+   */
+  void
+  add_syn_id( const synindex synid ) override
+  {
+    if ( not supports_syn_id( synid ) )
```

Comment:
I am a bit sceptical about this change. I understand that it fixes the mentioned issue. But why not using a `Set` instead of vector, if we don't allow multiple insertions of the same element, and the second question, why does it event happen that the same element will be inserted more than once.

Reply:
Will be fixed in a follow-up PR by @med-ayssar.

Reply:
Please refer to #2507